### PR TITLE
Add support for boolean columns to Arrow FSI

### PR DESCRIPTION
**Pull request by fexolm**

### CODE REVIEW

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp [hunk 1]
```diff
@@ -105,138 +100,160 @@ class ArrowForeignStorageBase : public PersistentForeignStorageInterface {
...
-                      dataT[j] = val;
-                    }
-                  }
-                }
-              });
+      tbb::blocked_range<size_t>(0, arr_col_chunked_array->num_chunks()),
+      [&](const tbb::blocked_range<size_t>& r) {
+        for (size_t c = r.begin(); c != r.end(); ++c) {
+          size_t offset = 0;
+          for (int i = 0; i < c; i++) {
```
Comment (paul-aiyedun) [inline DataMgr/ForeignStorage/ArrowForeignStorage.cpp:200]: Update `i` to `size_t` in order to avoid compiler warning.

**Review by paul-aiyedun — COMMENTED** [review]
Overall approach looks good to me. However, I had a question regarding the change in null value setting for non-boolean arrays, and there might be an issue with how boolean arrays with no null values are handled.

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp [hunk 2]
```diff
@@ -105,138 +100,160 @@ class ArrowForeignStorageBase : public PersistentForeignStorageInterface {
...
-                                                : std::numeric_limits<T>::max();
+  if ((!std::is_same_v<T, bool>)&&(arr_col_chunked_array->null_count() == 0)) {
+    // for boolean columns we still need to convert bitmaps to array
+    return arr_col_chunked_array;
+  }
+
+  auto null_value = get_null_value<T>();
+
+  auto resultBuf =
+      arrow::AllocateBuffer(sizeof(T) * arr_col_chunked_array->length()).ValueOrDie();
```
[the remark this answers is not in the document]
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp:192]: Yes, there is an issue, that we are modifying the array, allocated in shared memory. So, some other process may somehow rely on null values. Or maybe the arrow could decide to allocate the array in read-only memory in future updates. Anyway, copying the array here would enable zero-copy fetch in `read` function, so we do only one allocation anyway.

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp [hunk 3]
```diff
@@ -105,138 +100,160 @@ class ArrowForeignStorageBase : public PersistentForeignStorageInterface {
...
+          if (chunk->null_count() == chunk->length()) {
+            std::fill(resWithOffset, resWithOffset + chunk->length(), null_value);
+            continue;
+          }
+
+          auto chunkData = reinterpret_cast<const T*>(chunk->data()->buffers[1]->data());
+
+          if (chunk->null_count() == 0) {
+            std::copy(chunkData, chunkData + chunk->length(), resWithOffset);
+            continue;
```
[the remark this answers is not in the document]
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp:216]: Yeah, good point, I'll check it out

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp [hunk 4]
```diff
@@ -497,60 +508,50 @@ void ArrowForeignStorageBase::read(const ChunkKey& chunk_key,
...
-          varlen_offset += data[(sz / sizeof(uint32_t)) - 1];
-        }
-      } else {
-        auto fixed_type = dynamic_cast<arrow::FixedWidthType*>(array_data->type.get());
-        if (fixed_type) {
-          std::memcpy(
-              dest,
-              bp->data() + (array_data->offset + offset) * (fixed_type->bit_width() / 8),
-              sz = size * (fixed_type->bit_width() / 8));
+    CHECK(bp);
```
[the remark this answers is not in the document]
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp:511]: Only nulls case is now handled with replaceNullValues function. So, we always expect an allocated buffer here.

File: Tests/ArrowCsvForeignStorageTest.cpp [hunk 5]
```diff
@@ -499,6 +502,10 @@ TEST(NullValuesTest, NullFullColumn) {
   CHECK_EQ(0,
            v<int64_t>(run_simple_agg(
                "SELECT COUNT(fp8) FROM fsi_nulls_full WHERE fp8 IS NOT NULL;")));
+
+  CHECK_EQ(0,
+           v<int64_t>(run_simple_agg(
+               "SELECT COUNT(fp8) FROM fsi_nulls_full WHERE b8 IS NOT NULL;")));
 }
```
[the remark this answers is not in the document]
Comment (fexolm) [reply Tests/ArrowCsvForeignStorageTest.cpp:509]: Yeah, will do, thanks

**Comment by fexolm** [comment]
@paul-aiyedun I fixed the comments, PTAL.

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp [hunk 6]
```diff
@@ -263,18 +279,16 @@ int64_t ArrowForeignStorageBase::makeFragment(
...
+      if (buffers.size() <= 2) {
         throw std::runtime_error(
-            "Importing varialbe length arrow array as fixed length column");
+            "Importing fixed length arrow array as variable length column");
       }
+      auto offsets_buffer = reinterpret_cast<const uint32_t*>(buffers[1]->data());
+      varlen += offsets_buffer[offset + size] - offsets_buffer[offset];
+    } else if (buffers.size() != 2) {
+      throw std::runtime_error(
+          "Importing varialbe length arrow array as fixed length column");
```
[the remark this answers is not in the document]
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp]: fixed

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp [hunk 7]
```diff
@@ -105,138 +100,160 @@ class ArrowForeignStorageBase : public PersistentForeignStorageInterface {
...
+  if ((!std::is_same_v<T, bool>)&&(arr_col_chunked_array->null_count() == 0)) {
+    // for boolean columns we still need to convert bitmaps to array
+    return arr_col_chunked_array;
+  }
+
+  auto null_value = get_null_value<T>();
+
+  auto resultBuf =
+      arrow::AllocateBuffer(sizeof(T) * arr_col_chunked_array->length()).ValueOrDie();
+  auto resultData = reinterpret_cast<T*>(resultBuf->mutable_data());
```
[the remark this answers is not in the document]
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp]: fixed

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp]: fixed

File: Tests/ArrowCsvForeignStorageTest.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (fexolm) [reply Tests/ArrowCsvForeignStorageTest.cpp:509]: done

File: DataMgr/ForeignStorage/ArrowForeignStorage.cpp (hunk 1 above)
Comment (fexolm) [reply DataMgr/ForeignStorage/ArrowForeignStorage.cpp]: fixed

**Review by paul-aiyedun — APPROVED** [review]
Looks good to me.

**Comment by alexbaden** [comment]
I'm seeing an issue in this PR when running the tests:

```
2021-07-19T21:19:26.030229 I 1322788 0 QueryRunner.cpp:785 stdlog runSelectQuery 96 47 omnisci admin 909- {"query_str"} {"SELECT val2 FROM fsi_nulls_boolean order by id;"}
2021-07-19T21:19:26.030323 1 1322788 QueryRunner.cpp:785 stacked_times runSelectQuery 96 total time 47 ms
  processImpl 140567851819008 - total time 3 ms
2021-07-19T21:19:26.030353 1 1322788 0 QueryRunner.cpp:484 jsonlog { "name": "CPU Memory Info", "TotalMB": 64291, "FreeMB": 52530.7, "ProcessMB": 436.234, "VirtualMB": 111888, "ProcessPlusSwapMB": 211.551, "ProcessSharedMB": 224.684, "FragmentationPercent": 67 }
2021-07-19T21:19:26.030901 F 1322788 0 ArrowCsvForeignStorageTest.cpp:67 Check failed: exp == v<T>(crt_row[0]) (1 == 0) 
2021-07-19T21:19:26.031351 I 1322788 0 Calcite.cpp:649 Shutting down Calcite server
2021-07-19T21:19:26.033834 I 1322788 0 Calcite.cpp:661 shut down Calcite
shutdownSubsystemsAndExit: Interrupt signal (6) received.
```
the issue only seems to happen intermittently on our build machines, but I can reproduce it using Arrow 4.0 every time . Thoughts?

**Comment by fexolm** [comment]
Hmm, I'll take a look, thanks.

**Comment by fexolm** [comment]
@alexbaden Unfortunately, I was unable to reproduce the issue. Could you please share, how you build and execute tests?

**Comment by fexolm** [comment]
It looks like this pr is still not in master. @alexbaden @paul-aiyedun could you please take a look?

**Comment by fexolm** [comment]
Hey @mattpulver, I found a problem with no-nulls path, but it shouldn't really affect nulls test, I also rebased the branch on current master, so can you please rerun the CI to check how it works.
I couldn't really reproduce the issue even when compiling with clang.

**Comment by mattpulver** [comment]
@fexolm Looks like the recent updates fixed the previous clang error. We're reviewing it now on our internal repo w/ @paul-aiyedun's help.

**Comment by mattpulver** [comment]
Merged: https://github.com/omnisci/omniscidb/commit/0517d99b467806f6af7b4c969e351368a667497d

**Pull request closed without merging**